src/util/xeval/builtin_cast.rs

```diff
@@ -0,0 +1,335 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use tipb::expression::Expr;
+use util::codec::datum::Datum;
+use util::codec::mysql::{Decimal, Time, Duration, DEFAULT_FSP};
+use super::{Evaluator, EvalContext, Result, Error};
+
+fn invalid_type_error(datum: &Datum) -> Result<Datum> {
+    Err(Error::Eval(format!("invalid expr type: {:?}, expect to be int", datum)))
+}
+
+fn check_datum_int(datum: &Datum) -> Result<()> {
+    match *datum {
+        Datum::I64(_) | Datum::U64(_) => Ok(()),
+        _ => Err(Error::Eval(format!("invalid expr type: {:?}, expect to be int", datum))),
+    }
+}
+
+impl Evaluator {
+    pub fn cast_int_as_int(&mut self, ctx: &EvalContext, expr: &Expr) -> Result<Datum> {
+        let child = try!(self.get_one_child(expr));
+        let d = try!(self.eval(ctx, child));
+        match d {
+            Datum::I64(_) => Ok(d),
+            Datum::U64(_) => Ok(Datum::I64(d.i64())),
+            _ => invalid_type_error(&d),
+        }
+    }
+
+    pub fn cast_int_as_real(&mut self, ctx: &EvalContext, expr: &Expr) -> Result<Datum> {
+        let child = try!(self.get_one_child(expr));
+        let d = try!(self.eval(ctx, child));
+        match d {
+            Datum::I64(i) => Ok(Datum::F64(i as f64)),
+            Datum::U64(u) => Ok(Datum::F64(u as f64)),
+            _ => invalid_type_error(&d),
+        }
+    }
+
+    pub fn cast_int_as_string(&mut self, ctx: &EvalContext, expr: &Expr) -> Result<Datum> {
+        let child = try!(self.get_one_child(expr));
+        let d = try!(self.eval(ctx, child));
+        match d {
+            Datum::I64(_) | Datum::U64(_) => {}
+            _ => return invalid_type_error(&d),
+        }
+        let s = try!(d.into_string());
+        Ok(Datum::Bytes(s.into_bytes()))
+    }
+
+    pub fn cast_int_as_decimal(&mut self, ctx: &EvalContext, expr: &Expr) -> Result<Datum> {
+        let child = try!(self.get_one_child(expr));
+        let datum = try!(self.eval(ctx, child));
+        let decimal = match datum {
+            Datum::I64(_) => Decimal::from(datum.i64()),
+            Datum::U64(_) => Decimal::from(datum.u64()),
+            _ => return invalid_type_error(&datum),
+        };
+        Ok(Datum::Dec(decimal))
+    }
+
+    pub fn cast_int_as_time(&mut self, ctx: &EvalContext, expr: &Expr) -> Result<Datum> {
+        let child = try!(self.get_one_child(expr));
+        let d = try!(self.eval(ctx, child));
+        try!(check_datum_int(&d));
```

Comment:
It seems `Datum::U64` couldn't been converted to `Datum::Time` in tidb `https://github.com/pingcap/tidb/blob/master/util/types/datum.go#L1013`